quinn-proto/src/packet.rs

```diff
@@ -450,38 +427,65 @@ impl PartialEncode {
     }
 }
 
-pub enum InvariantHeader {
+pub enum PlainHeader {
+    Initial {
+        dst_cid: ConnectionId,
+        src_cid: ConnectionId,
+        token: Bytes,
+        len: u64,
+    },
     Long {
-        first: u8,
-        version: u32,
+        ty: LongType,
+        dst_cid: ConnectionId,
+        src_cid: ConnectionId,
+        len: u64,
+    },
+    Retry {
         dst_cid: ConnectionId,
         src_cid: ConnectionId,
+        orig_dst_cid: ConnectionId,
     },
     Short {
         first: u8,
+        spin: bool,
         dst_cid: ConnectionId,
     },
+    VersionNegotiate {
+        random: u8,
+        dst_cid: ConnectionId,
+        src_cid: ConnectionId,
+    },
 }
 
-impl InvariantHeader {
+impl PlainHeader {
     fn dst_cid(&self) -> ConnectionId {
-        use self::InvariantHeader::*;
+        use self::PlainHeader::*;
         match self {
+            Initial { dst_cid, .. } => *dst_cid,
             Long { dst_cid, .. } => *dst_cid,
+            Retry { dst_cid, .. } => *dst_cid,
             Short { dst_cid, .. } => *dst_cid,
+            VersionNegotiate { dst_cid, .. } => *dst_cid,
         }
     }
 
     fn decode<R: Buf>(buf: &mut R, local_cid_len: usize) -> Result<Self, PacketDecodeError> {
```

Comment:
To get zero-copy tokens at this early stage, this function will need to become monomorphic, taking a `(&mut? )BytesMut` and consuming the portion of it that was read so that the `Bytes` can be pulled out. This will require changing the current `HeaderCrypto` API to take `first: &mut u8, pn: &mut [u8]` or similar instead of a single contiguous slice, and the inclusion of a `first` field in `PlainHeader::Initial`/`Long`.

Reply:
Or maybe `HeaderCrypto` should include parsing, so it can take value/shared reference arguments and return flags/pn?

Reply:
To be clear, I don't consider this a blocker, but it'd be neat.

Reply:
To be honest, all of those options sound like a whole lot of complexity just to avoid a single allocation. I think we should probably do some benchmarks before we do those kinds of optimization.